contrib/zenith/pagestore_prefetch.c

```diff
@@ -0,0 +1,440 @@
+/*-------------------------------------------------------------------------
+ *
+ * pagestore_prefetch.c
+ *
+ *
+ * Portions Copyright (c) 1996-2021, PostgreSQL Global Development Group
+ * Portions Copyright (c) 1994, Regents of the University of California
+ *
+ *
+ * IDENTIFICATION
+ *	  contrib/zenith/pagestore_prefetch.c
+ *
+ *-------------------------------------------------------------------------
+ */
+#include "postgres.h"
+#include "miscadmin.h"
+#include "pgstat.h"
+#include "pagestore_client.h"
+#include "postmaster/bgworker.h"
+#include "storage/relfilenode.h"
+#include "storage/buf_internals.h"
+#include "storage/latch.h"
+#include "storage/ipc.h"
+#include "storage/lwlock.h"
+#include "utils/dynahash.h"
+#include "utils/guc.h"
+#include "utils/timestamp.h"
+#include "pagestore_client.h"
+
+/*
+ * Prefetch is using fixed size ring buffer and postgres hash table. Both are located in shared memory.
+ * Backends are adding new entries to the ring buffer and prefetch background worker is taken them from ring buffer.
+ * When smgrprefetch() is called, we try to locate entry in hash table. If it is not present yet, then we add entry to ring buffer.
+ * We do no check overflow of ring buffer: if entry was not yet proceeded by prefetch worker, then we just replace it because
+ * this request is considered as deteriorated.
+ *
+ * When smgrread is called, we first try to locate entry in prefetch hash table. If request is present in hash table, then it can be in
+ * one of three states:
+ * 1. Not yet proceeded (sent to page server)
+ * 2. Sent to page server, but response is not yet received.
+ * 3. Proceeded: prefetched page is placed in ring buffer
+ * In the first case we just cancel prefetch request and let backend request page from page server itself.
+ * In the second case we wait until request is completed.
+ * And in the third case copy data from ring buffer to the destination and immediately return
```

Comment:
May I ask you to move those descriptions to their corresponding bullets? That'd be more comprehensible.

E.g.

```markdown
- First scenario (brief description)
  Profound description of this outcome and its consequences.
  
- Second scenario (brief description)
  ...
  
- ...
```